en/news/_posts/2019-12-06-separation-of-positional-and-keyword-arguments-in-ruby-3-0.md

```diff
@@ -0,0 +1,377 @@
+```
+
+If a method accepts both optional and keyword arguments, the Hash object that has both Symbol keys and non-Symbol keys was "split" in Ruby 2.6.  In Ruby 2.7, however, both are accepted as keywords  because non-Symbol keys are allowed.
+
+```
+def bar(x=1, **kwargs)
+  p [x, kwargs]
+end
+
+bar("key" => 42, :sym => 43)
+  #=> Ruby 2.6: [{"key"=>42}, {:sym=>43}]
+  #=> Ruby 2.7: [1, {"key"=>42, :sym=>43}]
+
+# Use braces to keep the behavior
+bar({"key" => 42}, :sym => 43)
+  #=> Ruby 2.6 and 2.7: [{"key"=>42}, {:sym=>43}]
+```
+
+Ruby 2.7 will still split hashes with a warning if passing a Hash or keyword arguments with both Symbol and non-Symbol keys to a method that accepts explicit keywords but no keyword rest argument (`**kwargs`).  This splitting will be removed in Ruby 3, and an `ArgumentError` will be raised.
+
+```
+def bar(x=1, sym: nil)
+  p [x, sym]
+end
+
+bar("key" => 42, :sym => 43)
+# Ruby 2.6 and 2.7: => [{"key"=>42}, 43]
+# Ruby 2.7: warning: The last argument is split into positional and keyword parameters
+#           warning: for `bar' defined here
+# Ruby 3.0: ArgumentError
+```
+
+### 2. Double splat with an empty hash (`**{}`)  passes no arguments
+
+In Ruby 2.6 or before, passing `**empty_hash` passes an empty Hash as a positional argument.  In Ruby 2.7 or later, it passes no arguments.
+
+```
+def foo(*args)
+  args
+end
+
+empty_hash = {}
+foo(**empty_hash)
+  #=> Ruby 2.6 or before: [{}]
+  #=> Ruby 2.7 or later: []
+```
+
+Note that `foo(**{})` passes nothing in both Ruby 2.6 and 2.7.  In Ruby 2.6 and before, `**{}` is removed by the parser, and in Ruby 2.7 and above, it is treated the same as `**empty_hash`, allowing for an easy way to pass no keyword arguments to a method.
+
+In Ruby 2.7, when calling a method with an insufficient number of required positional arguments, `foo(**empty_hash)` passes an empty hash with a warning emitted, for compatibility with Ruby 2.6.  This behavior will be removed in 3.0.
+
+```
+def foo(x)
+  x
+end
+
+empty_hash = {}
+foo(**empty_hash)
+  #=> Ruby 2.6 or before: {}
+  #=> Ruby 2.7: warning: The keyword argument is passed as the last hash parameter
+  #             warning: for `foo' defined here
+  #=> Ruby 3.0: ArgumentError: wrong number of arguments
+```
+
+### 3. The no-keyword-arguments syntax (`**nil`) is introduced
+
+`**nil` is allowed in method definitions to explicitly mark that the method accepts no keyword arguments. Calling such a method with keyword arguments will result in an `ArgumentError`.  (This is not an incompatibility but a new feature.)
+
+```
+def foo(*args, **nil)
+end
+
+foo(k: 1)
+  #=> Ruby 2.7 or later: no keywords accepted (ArgumentError)
+```
+
+This is useful to make it explicit that the method does not accept keyword arguments.  Otherwise, the keywords are absorbed in the rest argument in the above example.  If you want to extend the method to accept a keyword argument in future, it may break the compatibility.
```

Comment:
```suggestion
This is useful to make it explicit that the method does not accept keyword arguments.  Otherwise, the keywords are absorbed in the rest argument in the above example.  If you extend a method to accept keyword arguments, the method may have incompatibility as follows:
```